### PR TITLE
 Enable pretty import paths for local imports (from repo) in brigade.js

### PR DESCRIPTION
**Pull request by radu-matei**
**What this PR does / why we need it**:
This PR enabled pretty import paths for local JavaScript modules from the same repo as the Brigade project, without the user having to know about the destination in the Brigade worker pod of the source code.

Specifically, it intercepts all local path imports in `brigade.js` and overrides them to the absolute path in the worker pod (`/vcs/<rest-of-relative-path>`).

For more context, see https://github.com/Azure/brigade/issues/450#issuecomment-458365387.

This PR does _not_ contain the docs for this feature, as #450 takes the larger task of documenting that in detail.

- [x] this PR contains unit tests
- [x] this PR contains documentation

### CODE REVIEW

**Comment by dgkanatsios** [comment]
Manually tested with your repo [here](https://github.com/radu-matei/brigade-javascript-deps), works like a charm. Only one code question, else LGTM.

File: brigade-worker/src/require.ts [hunk 1]
```diff
@@ -16,10 +18,20 @@ class Pkg {
...
+export function getOverriddenPackage(pkg: string): string {
   for (let p of pkgOverrides) {
     if (p.name == pkg) {
-      return require(p.override);
+      return p.override;
+    }
+
+    // we want to intercept the loading of relative modules in the repo
+    // and and override them with the absolute path inside the worker pod
+    if ((!path.isAbsolute(pkg)) && (pkg.includes("./"))) {
```
Comment (dgkanatsios) [inline brigade-worker/src/require.ts]: shouldn't it be `path.startsWith`?

Comment (radu-matei) [reply brigade-worker/src/require.ts]: Thanks for catching that, updating.

Comment (radu-matei) [reply brigade-worker/src/require.ts]: Although on second thought, here's a scenario I haven't tested yet - 
In the context of #766, with `brigade.js` is in a custom location, say in a directory called `brigade`, and I want to load a module from `../js/my-module`, both our assumptions fail:

- the path does not start with `./`;
- the result of that path manipulation is not correct - it becomes `"./vcs/js/my-module"`, which is not the correct absolute path.

Comment (dgkanatsios) [reply brigade-worker/src/require.ts]: hm.... trying to think of something else as well. brigade.js references fileA.js, which references fileB.js. Do we need to do any changes in fileA or fileB?

**Comment by radu-matei** [comment]
I honestly don't have a preference for that.
I was thinking, because I'd like the doc to cover importing both local and NPM packages, to have it as a separate PR, in a more tutorial style.

WDYT?

**Comment by vdice** [comment]
Separate PR makes sense, especially w/ the extended examples you've mentioned.  👍

**Comment by radu-matei** [comment]
@vdice, in its current state, this PR does not allow importing local modules when `brigade.js` is in a relative directory  (context of #766) - mainly because it involves resolving two levels of relative paths:
- the path of `brigade.js` relative to the repo root
- the path of every imported local module relative to _both_ the repo root and `brigade.js`

Given that having `brigade.js` in a custom location is a corner case, do we want to merge this PR in its current form, mark it as experimental, and clearly state the described scenario is not supported?

**Comment by vdice** [comment]
@radu-matei Yes, I think that sounds fair for now.  Perhaps the disclaimer can mention that only local modules existing on same directory level as the `brigade.js`file are currently supported -- then, we'd be supporting the custom brigade.js location as well (or could, with an updated bit of logic to pull in custom path prefix if set).

File: docs/topics/dependencies.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+[brigade:k8s] Creating PVC named brigade-worker-01cexrvrs08shcev26961cwd6n
+is 13 thirteen? true
+[brigade:app] after: default event handler fired
+[brigade:app] beforeExit(2): destroying storage
+[brigade:k8s] Destroying PVC named brigade-worker-01cexrvrs08shcev26961cwd6n
+```
+
+Notes:
+
+- the dependencies should point the exact version (and not use the tilde ~ and caret ^ to indicate semver compatible versions)
```
Comment (vdice) [inline docs/topics/dependencies.md]: Is `should` the right word here -- or is `must` the right one?  (Basically, will Brigade's dep fetching break down if an exact version is _not_ supplied?)

Comment (vdice) [reply docs/topics/dependencies.md]: typo: missing `to` in  `...point the exact version...`

File: docs/topics/dependencies.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+- when adding a custom dependency using `brigade.json`, `yarn` will add it side-by-side with [the worker's 
+dependencies](../../brigade-worker/package.json) - this means that the process will fail if a dependency that conflicts with one of the 
+worker's dependency is added. However, already existing dependencies (such as `@kubernetes/client-node`, `ulid` or `chai`)
```
Comment (vdice) [inline docs/topics/dependencies.md]: typo: pluralization needed, should read `... worker's dependencies...`

Comment (vdice) [reply docs/topics/dependencies.md]: Curious -- what does the failure look like to the user?  Will it be obvious?

File: docs/topics/dependencies.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+- as the Brigade worker is capable of modifying the state of the cluster, be mindful 
+of any external dependencies added through `brigade.json`
+
+- for now, the only part of `brigade.json` used is the `dependencies` object - it means the rest of the file 
+can be used to pass additional information to `brigade.js`, such as environment data - however, as the `brigade.json` 
+file is passed in the source control repository, information passed there will be available to anyone with access to the repository.
+
+- all dependencies added in `brigade.json` are dynamically installed on every Brigade build - this means if the dependencies added 
+are large, and the build frequency is high for a particular project, it might make sense to make a pre-built Docker image that 
+already contains the dependencies, as described in the sections below.
```
Comment (vdice) [inline docs/topics/dependencies.md]: update needed here -- no longer `described in the sections below`, but back in the `workers.md` doc that we'd linked to above...

File: docs/topics/dependencies.md (hunk 2 above)
Comment (radu-matei) [reply docs/topics/dependencies.md]: It's a _must_, thanks for catching this!

File: docs/topics/dependencies.md (hunk 3 above)
Comment (radu-matei) [reply docs/topics/dependencies.md]: Hm, that's a very good point - let me open a new issue for that. Thanks!

**Comment by radu-matei** [comment]
Thanks a lot for the reviews!

**Pull request merged**